File: lib/game.rb
class Game
  def self.play
    Game.new.tap(&:play)
  end

  def initialize
    @valid_choices = [
      :help,
      :exit,
      :north,
      :south,
      :east,
      :west,
      :display_map,
      :look_around,
      :pick_up,
      :use,
      :print_inventory,
      :talk,
      :take,
      :move
    ]

    @player = Player.new
    @playing = true
    @map = Map.new("Liz's Great Adventure", rooms, @player)
  end

  def play
		puts "Welcome to #{@map.title}"
    puts "What would you like to do? (Enter 'help' to see a list of commands)"
    parse_choice(gets.chomp)
    while @playing
      break if !@playing
      puts "What now?"
      choice = gets.chomp
      parse_choice(choice)
    end
  end

  private

  # take item from an NPC
  def take(item_name)
    if current_room.npc && current_room.npc.has_item(item_name)
      # inventory should be an array and not an InventoryItem object
      item = current_room.npc.inventory.name == item_name ? current_room.npc.inventory : nil
      # how to remove inventory item from npc?
      # not sure inventory is being created correctly
      npc_item = current_room.npc.inventory.pop(item)
      @player.add_to_inventory(npc_item)
    else
      puts "Sorry, that item isn't here."
    end
  end

  # take item from a Room
  def pick_up(item_name)
    if current_room.has_item(item_name)
      item = current_room.items.find { |thing| thing.name == item_name }
      @player.add_to_inventory(item)
      current_room.remove_one(item)
    else
      puts "Sorry, that item is not in this room. Try again."
    end
  end

	# use an item from your inventory
  def use(item_name)
    # TODO: figure out how to pass around entire item object to access effects anywhere
    if @player.has_item(item_name) && current_room.npc && current_room.npc.has_item(item_name)
      effect = current_room.npc.inventory.effects
      puts effect
      @player.remove_from_inventory(item_name)
      # TODO: eventually remove from NPC inventory & change ownership of item
    elsif @player.has_item(item_name) && current_room.has_item(item_name)
      effect = current_room.items.find { |i| i.name == item_name }.effects
      puts effect
      @player.remove_from_inventory(item_name)
      # TODO: eventually remove from Room inventory & change ownership of item
    else
      puts "Sorry, that item is not in your inventory. Did you pick it up or try taking it from someone?"
    end
  end

	# talk to an NPC
  def talk
    if current_room.npc
      puts current_room.npc.dialogue[:default]
    else
      puts "Sorry, no one else is in here!"
    end
  end

	# print your inventory to the console
  def print_inventory
    @player.print_inventory
  end

  #TODO: clean this method up like whoa
  def parse_choice(choice)
    new_choice = choice.split.join("_")
    valid_choice = @valid_choices.select { |entry| entry == new_choice.to_sym }.first
    if choice.include?('move') && valid_move(choice)
      move(choice.split(' ').last.to_sym)
    elsif choice.include?('pick up')
      item = choice.split.select{ |item| item != 'pick' && item != 'up' }.join(' ')
      pick_up(item)
    elsif choice.include?('take')
      item = choice.split.select{ |item| item != 'take' }.join(' ')
      take(item)
    elsif choice.include?('use')
      item = choice.split.select{ |item| item != 'use' }.join(' ')
      use(item)
    elsif valid_choice != nil
      self.send valid_choice
    else
      check_validity(new_choice.to_sym)
    end
  end

  # check that move is in a valid direction
  def valid_move(user_input)
    user_input.split(' ').first == 'move' &&
      ['north','south', 'east', 'west'].include?(user_input.split(' ').last)
  end

	# see all the rooms
  def display_map
    @map.display_map
  end

  def current_room
    rooms.select { |room| room[0] == @player.location }.first
  end

  def look_around
    puts current_room.description
    if current_room.has_items
      puts "This room contains #{current_room.item_list}"
    end
    puts current_room.npc.name + " is here too." if current_room.npc
  end

  def move(direction)
    if @map.valid_directions(@player.location)[direction]
      case direction
      when :north
        @player.location.y += 1
        puts "You have moved north! Your new location is #{@player.location.x}, #{@player.location.y}"
      when :south
        @player.location.y -= 1
        puts "You have moved south! Your new location is #{@player.location.x}, #{@player.location.y}"
      when :west
        @player.location.x -= 1
        puts "You have moved west! Your new location is #{@player.location.x}, #{@player.location.y}"
      when :east
        @player.location.x += 1
        puts "You have moved east! Your new location is #{@player.location.x}, #{@player.location.y}"
      else
        puts "Sorry, you can't move there!"
      end
    else
      puts "Sorry, you can't move there!"
    end
  end

  def help
    puts <<-HEREDOC
      exit: exit the game
      north, south, east, west: move in this direction
      look around: see a description of the current room
      pick up _item_: add the item to your inventory
      take _item_: take an item from an NPC
      use _item_: use an item in your inventory
      talk: talk to an NPC
      display map: look at map
      print inventory: show current player inventory
    HEREDOC
  end

  def exit
    puts "Bye, #{@player.name}! Thanks for playing!"
    abort
  end

  def rooms
    [
      Room.new(
        0, 0,
        "Unicorn Room",
        "This room contains a rare and glorious unicorn. It's amazing.",
        [InventoryItem.new("a jar of unicorn farts", 1, nil, "BOOM. Unicorn farts are powerful! You are now very sparkly.")],
        NPC.new(
          "Unicorn Doctor",
          InventoryItem.new("vial of unicorn blood", 1, self, "The amazing unicorn blood has made you INVINCIBLE!"),
          {
            default: "Hi I'm a unicorn doctor. It's pretty cool. I have a vial of unicorn blood, do you want it?"
          }
        ),
      ),
      Room.new(
        0, 1,
        "Bear Room",
        "HOLY CRAP THERE'S A BEAR IN THIS ROOM.",
        [InventoryItem.new("a canister of bear repellant", 1, nil, "You are now safe from the bear! Stop stressin'")],
        nil
      ),
      Room.new(
        0, 2,
        "Cool Stuff Room",
        "This room's pretty cool, nbd",
        [nil],
        nil
      ),
      Room.new(
        1, 0,
        "Crappy Stuff Room",
        "Everything in this room stinks like garbage.",
        [InventoryItem.new("a garbage bomb", 3, nil, "Bad move, now you stink like garbage. But so does everything else.")],
        nil
      ),
      Room.new(
        1, 1,
        "Starting Out Room",
        "This is a room to start out in. Nothing to see here.",
        [nil],
        nil
      ),
      Room.new(
        1, 2,
        "Cute Puppy Room",
        "OMG this room is FULL. OF. PUPPIES. So many puppies!",
        [InventoryItem.new("a puppy", 10, nil, "You pet the heck out of the puppy. Look at his waggy tail! LOOK AT IT!! You are hypnotized.")],
        nil
      ),
      Room.new(
        2, 0,
        "Sandwich and Chips Room",
        "Yum, there's a sandwich and some chips in here!",
        [
          InventoryItem.new("a sandwich", 1, nil, "Yum, that was a good sandwich. It was made of whatever your favorite sandwich is."),
          InventoryItem.new("a bag of chips", 1, nil, "That was a delightful bag of chips! Crunchy as heck with perfect salt.")
        ],
        nil
      ),
      Room.new(
        2, 1,
        "Home Alone Room",
        "This room's got nothing in it. You're allll aloooone.",
        [nil],
        nil
      ),
      Room.new(
        2, 2,
        "Dank Meme Room",
        "This room is nothing but sweet memes.",
        [
          InventoryItem.new("a pic of Hillary Clinton texting", 1, nil, "You laugh yourself to sleep because memes are so funny right"),
          InventoryItem.new("a pic of a dog getting hit in the face with a frisbee", 1, nil, "Man that frisbee dog is hilarious, isn't he? Memes are the best.")
        ],
        nil
      )
    ]
  end

  # checks that Player input is valid
  def check_validity(choice)
    valid_options = ['pick up', 'use', 'display inventory', 'move']
    unless @valid_choices.include?(choice.to_sym) || valid_options.include?(choice)
      puts "That is not a valid choice. Try again."
    end
  end
end

File: src/game.rs
use player::Player;
use map::Map;
use room::Room;
use inventory_item::InventoryItem;
use std::io;

extern crate regex;
use Chomp;

// where most player console interactions and game loop will be defined
#[derive(Debug, Eq, PartialEq)]
pub struct Game {
    player: Player,
    map: Map,
    pub playing: bool,
}

impl Game {
    fn start(player: Player, map: Map, playing: bool) {
        let mut game = Game::new(player, map, playing);
        game.play();
    }

    pub fn new(player: Player, map: Map, playing: bool) -> Game {
        println!("Hi {}Welcome to {}", player.name, map.title);
		println!("What would you like to do? (Enter 'help' to see a list of commands)");

        Game {
            player: player,
            map: map,
            playing: playing,
        }
    }

    pub fn play(&mut self) {
        println!("What would you like to do?");
        self.parse_choice();
    }

    fn current_room(&self) -> &Room {
        self.map.rooms.get(&self.player.location).expect("BUG: The player's location must exist in the map")
    }

    fn current_room_mut(&mut self) -> &mut Room {
        self.map.rooms.get_mut(&self.player.location).expect("BUG: The player's location must exist in the map")
    }

    fn player(&self) -> &Player {
        &self.player
    }

    fn player_mut(&mut self) -> &mut Player {
        &mut self.player
    }

    fn parse_choice(&mut self) {
        let mut user_input = String::new();

        io::stdin()
            .read_line(&mut user_input)
            .expect("Could not read line");

        let user_input = user_input.chomp();

        if user_input == "look around" {
            self.look_around();
        } else if user_input == "talk" {
            self.talk();
        } else if user_input == "display map" {
            self.map.display_map();
        } else if user_input == "print inventory" {
            self.player.print_inventory();
        } else if user_input == "exit" {
            println!("Thanks for playing!");
            self.playing = false;
        } else if user_input == "help" {
            println!("exit: exit the game");
            println!("move north, south, east, west: move in this direction");
            println!("look around: see a description of the current room");
            println!("pick up _item_: add the item to your inventory");
            println!("take _item_: take an item from an NPC");
            println!("use _item_: use an item in your inventory");
            println!("talk: talk to an NPC");
            println!("display map: look at map");
            println!("print inventory: show current player inventory");
        } else if let Some(captures) = regex("(?i)^pick up (?P<thing>.*)").captures(user_input) {
            self.pick_up(captures.name("thing").expect("unexpected optional capture"));
        } else if let Some(captures) = regex("(?i)^take (?P<thing>.*)").captures(user_input) {
            self.take(captures.name("thing").expect("unexpected optional capture"));
        } else if let Some(captures) = regex("(?i)^use (?P<thing>.*)").captures(user_input) {
            self.use_item(captures.name("thing").expect("unexpected optional capture"));
        } else if let Some(captures) = regex("(?i)^move (?P<direction>.*)").captures(user_input) {
            self.change_location(captures.name("direction").expect("unexpected optional capture"));
        } else {
            println!("{:?}", user_input);
            let split_input = user_input.split_whitespace();
            println!("{:?}", split_input.collect::<Vec<&str>>());
        }
    }

    // MOVES //

    fn string_to_inventory_item(&mut self, item: &str) -> Option<InventoryItem> {
        let items = self.current_room_mut().items_mut();

        items.iter()
            .position(|thing| item == thing.name)
            .map(|i| items.remove(i))
    }

    fn string_to_npc_item(&mut self, item_name: &str) -> Option<InventoryItem> {
        let npc_inventory = self.current_room_mut().npc_mut().inventory_mut();

        npc_inventory.iter()
            .position(|thing| item_name == thing.name)
            .map(|i| npc_inventory.remove(i))
    }

    fn string_to_player_item(&mut self, item_name: &str) -> Option<InventoryItem> {
        let player_inventory = self.player_mut().inventory_mut();

        player_inventory.iter()
            .position(|thing| item_name == thing.name)
            .map(|i| player_inventory.remove(i))
    }

    fn pick_up(&mut self, item_name: &str) {
        match self.string_to_inventory_item(item_name) {
            Some(item) => { self.player.add_to_inventory(item); }
            None => println!("Sorry, {} wasn't found in the current room", item_name)
        };
    }

    fn take(&mut self, item_name: &str) {
        match self.string_to_npc_item(item_name) {
            Some(item) => { self.player.add_to_inventory(item); }
            None => println!("Sorry, {} doesn't have {}.", self.current_room().npc().name, item_name)
        }
    }

    fn use_item(&mut self, item_name: &str) {
        match self.string_to_player_item(item_name) {
            Some(item) => { println!("{}", item.effects) }
            None => println!("Sorry, you don't have {} in your inventory.", item_name)
        }
    }

    fn change_location(&mut self, direction: &str) {
        let valid_directions = self.map.valid_directions(&self.player().location());

        if direction == "north" {
            if valid_directions.north {
                self.player_mut().location_mut().y += 1;
                println!("You have moved north.");
            } else {
                println!("You can not go north. Try a different direction.");
            }
        } else if direction == "south" {
            if valid_directions.south {
                self.player_mut().location_mut().y -= 1;
                println!("You have moved south.");
            } else {
                println!("You can not go south. Try a different direction.");
            }
        } else if direction == "west" {
            if valid_directions.west {
                println!("You have moved west.");
                self.player_mut().location_mut().x -= 1;
            } else {
                println!("You can not go west. Try a different direction.");
            }
        } else if direction == "east" {
            if valid_directions.east {
                println!("You have moved east.");
                self.player_mut().location_mut().x += 1;
            } else {
                println!("You can not go north. Try a different direction.");
            }
        } else {
            println!("That is not a valid direction. Try north, south, east, or west.");
        }
    }

    fn look_around(&self) {
        // display the current room's description
        println!("{}", &self.current_room().description);
        // if the room has any items, display information about them
        if !&self.current_room().items.is_empty() {
            println!("This room contains: {:?}", &self.current_room().items);
        }
        // display information about the room's NPC
        println!("{} is here too!", &self.current_room().npc.name);

        if !&self.current_room().npc().inventory.is_empty() {
            println!("{} has {:?}.", &self.current_room().npc.name, &self.current_room().npc.inventory)
        }
    }

    fn talk(&self) {
        println!("{}", &self.current_room().npc.dialogue);
    }
}

fn regex(s: &str) -> regex::Regex {
    regex::Regex::new(s).unwrap()
}
